File: LinkedList/python/middle_of_the_linked_list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def size(self,head):
        temp = head
        len = 0
        while temp!=None:
            len+=1
            temp = temp.next
        return len

    def middleNode(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        half = (self.size(head)//2)
        temp = head
        i = 0
        while i!=half:
            temp = temp.next
            i+=1

        return temp
        
# another approach . 
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def middleNode(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        slow = head
        first = head
        while first!=None and first.next!=None:
            slow = slow.next
            first = first.next.next

        return slow
        

File: LinkedList/typescript/middle_of_the_linked_list.ts
/**
 * Definition for singly-linked list.
*/
// class ListNode {
//     val: number
//     next: ListNode | null
//     constructor(val?: number, next?: ListNode | null) {
//        this.val = (val===undefined ? 0 : val)
//        this.next = (next===undefined ? null : next)
//      }
// }
// function size(head:ListNode|null):number{
//     let temp = head
//     let length:number = 0
//     while (temp!=null){
//         length++
//         temp = temp.next
//     }
//     return length
// }
// function middleNode(head: ListNode | null): ListNode | null {
//     const half = Math.trunc(size(head)/2)
//     let temp:ListNode|null = head
//     for(let i:number=0;i!=half;i++){
//         if(temp!=null){
//             temp = temp?.next
//         }
//     }
//     return temp
// };
// Another approach.
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
       this.val = (val===undefined ? 0 : val)
       this.next = (next===undefined ? null : next)
     }
}

function middleNode(head: ListNode | null): ListNode | null {
    
    let slow:ListNode|null = head
    let first:ListNode|null = head
    while (first!=null && first.next!=null && slow!=null){
        slow = slow.next
        first = first.next.next
    }
    return slow
};